ledger/src/shred.rs

```diff
@@ -775,100 +761,53 @@ impl Shredder {
-            let data_ptrs: Vec<_> = data_shred_batch
-                .iter()
-                .map(|data| &data.payload[..valid_data_len])
-                .collect();
-
-            // Create empty coding shreds, with correctly populated headers
-            let mut coding_shreds: Vec<_> = (0..num_coding)
-                .map(|i| {
-                    Shred::new_empty_coding(
-                        slot,
-                        start_index + i as u32,
-                        fec_set_index,
-                        num_data,
-                        num_coding,
-                        i, // position
-                        version,
-                    )
-                    .payload
-                })
-                .collect();
-
-            // Grab pointers for the coding blocks; these come after the two headers
-            let coding_block_offset = SIZE_OF_CODING_SHRED_HEADERS;
-            let mut coding_ptrs: Vec<_> = coding_shreds
-                .iter_mut()
-                .map(|buffer| &mut buffer[coding_block_offset..])
-                .collect();
-
-            // Create coding blocks
-            session
-                .encode(&data_ptrs, coding_ptrs.as_mut_slice())
-                .expect("Failed in erasure encode");
-
-            // append to the shred list
-            coding_shreds
-                .into_iter()
-                .enumerate()
-                .map(|(i, payload)| {
-                    let mut shred = Shred::new_empty_coding(
-                        slot,
-                        start_index + i as u32,
-                        start_index,
-                        num_data,
-                        num_coding,
-                        i,
-                        version,
-                    );
-                    shred.payload = payload;
-                    shred
-                })
-                .collect()
-        } else {
-            vec![]
-        }
-    }
-
-    fn calculate_num_coding_shreds(
-        num_data_shreds: usize,
-        fec_rate: f32,
-        max_coding_shreds: usize,
-    ) -> usize {
-        if num_data_shreds == 0 {
-            0
+    pub fn generate_coding_shreds(data: &[Shred], is_last_in_slot: bool) -> Vec<Shred> {
+        const PAYLOAD_ENCODE_SIZE: usize = SHRED_PAYLOAD_SIZE - SIZE_OF_CODING_SHRED_HEADERS;
+        let ShredCommonHeader {
+            slot,
+            index,
+            version,
+            fec_set_index,
+            ..
+        } = data.first().unwrap().common_header;
+        assert_eq!(fec_set_index, index);
+        assert!(data.iter().all(|shred| shred.common_header.slot == slot
+            && shred.common_header.version == version
+            && shred.common_header.fec_set_index == fec_set_index));
+        let num_data = data.len();
+        let num_coding = if is_last_in_slot {
+            (2 * MAX_DATA_SHREDS_PER_FEC_BLOCK as usize)
+                .saturating_sub(num_data)
+                .max(num_data)
```

Comment:
sanity check, `num_data` should always be `<= MAX_DATA_SHREDS_PER_FEC_BLOCK`,  due to chunking by the caller right? Should we add this as an assert?

Reply:
I do not think we need an `assert`, since this function may be called directly and larger batches are totally valid.

Reply:
yeah, was wondering if we wanted to ensure that `num_data + num_coding < 2 * MAX_DATA_SHREDS_PER_FEC_BLOCK`, but if we are planning for supporting larger batches as well, then that makes sense!.